### PR TITLE
Tipset and TipsetKeys property tests

### PR DESCRIPTION
**Pull request by connormullett**
**Summary of changes**
Changes introduced in this pull request:
- Add quickcheck tests for tipset and tipsetkeys using quickcheck

**Reference issue to close (if applicable)**

Closes 
part of #1881, affected tasks have been marked complete

**Other information and links**

### CODE REVIEW

File: blockchain/blocks/src/tipset.rs [hunk 1]
```diff
@@ -47,6 +47,15 @@ impl PartialEq for Tipset {
     }
 }
 
+impl quickcheck::Arbitrary for TipsetKeys {
+    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
+        let header = BlockHeader::arbitrary(g);
+        Self {
+            cids: vec![*header.cid()],
```
Comment (lemmih) [inline blockchain/blocks/src/tipset.rs:54]: You don't have to create a block header just to get a CID. Instead, just generate a vector of random CIDs.

File: blockchain/blocks/src/tipset.rs [hunk 2]
```diff
@@ -543,3 +552,27 @@ mod test {
...
+        let serialized = serde_json::to_string(&TipsetKeysJson(tipset_keys.clone())).unwrap();
+        let parsed: TipsetKeysJson = serde_json::from_str(&serialized).unwrap();
+        assert_eq!(tipset_keys, parsed.0);
+    }
+
+    #[quickcheck]
+    fn tipset_roundtrip(tipset: Tipset) {
+        let serialized = serde_json::to_string(&TipsetJsonRef(&tipset)).unwrap();
+        let parsed: TipsetJson = serde_json::from_str(&serialized).unwrap();
+        assert_eq!(Arc::new(tipset), parsed.0);
```
Comment (lemmih) [inline blockchain/blocks/src/tipset.rs:576]:
```suggestion
        assert_eq!(&tipset, parsed.0.as_ref());
```

**Review by lemmih — CHANGES_REQUESTED** [review]

File: blockchain/blocks/src/tipset.rs (hunk 1 above)
Comment (connormullett) [reply blockchain/blocks/src/tipset.rs:54]: 👍

File: blockchain/blocks/src/tipset.rs [hunk 3]
```diff
@@ -47,6 +47,16 @@ impl PartialEq for Tipset {
     }
 }
 
+impl quickcheck::Arbitrary for TipsetKeys {
+    fn arbitrary(g: &mut quickcheck::Gen) -> Self {
+        let cid = Cid::new_v1(
+            u64::arbitrary(g),
+            cid::multihash::Multihash::wrap(u64::arbitrary(g), &[u8::arbitrary(g)]).unwrap(),
+        );
+        Self { cids: vec![cid] }
```
Comment (lemmih) [inline blockchain/blocks/src/tipset.rs:56]: We also want to test what happens when you have lots of keys or zero keys.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: blockchain/blocks/src/tipset.rs (hunk 3 above)
Comment (lemmih) [reply blockchain/blocks/src/tipset.rs:56]: (You could implement this by creating a wrapper around `Cid`)

Comment (connormullett) [reply blockchain/blocks/src/tipset.rs:56]: Shouldn't that be part of a normal unit test? As far as I'm aware, these tests are just for checking if serialization/deserialization works both ways.

Comment (lemmih) [reply blockchain/blocks/src/tipset.rs:56]: I'm not sure what you mean. As the code is right now, you're just testing if serialization/de-serialization works when `TipsetKeys` contains a single `Cid`. We also want to know if it works when there are zero Cids or many Cids.

**Review by lemmih — APPROVED** [review]

**Review by jdjaustin — COMMENTED** [review]
Probably nit, but @LesnyRumcajs has been asking me to put the `struct` and `impl` blocks under each property test module (i.e., so there is only one `#[cfg(test)]` for all the property test code), and the code is definitely a lot cleaner that way.

**Comment by LesnyRumcajs** [comment]
Good point @jdjaustin! :)

**Comment by connormullett** [comment]
> Probably nit, but @LesnyRumcajs has been asking me to put the `struct` and `impl` blocks under each property test module (i.e., so there is only one `#[cfg(test)]` for all the property test code), and the code is definitely a lot cleaner that way.

Not a nitpick at all, this makes sense. Fixing now

**Review by jdjaustin — APPROVED** [review]
